### ISSUE TITLE
[PATCH] FTPClient does not fire current commands errback on connectionLost

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11434]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| orbitz reported|
|-|-|
|Trac ID|trac#1107|
|Type|defect|
|Created|2005-07-14 02:15:53Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1107 1107
type__defect defect
reporter__orbitz orbitz
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1121307353000000 1121307353000000
changetime__1175411415000000 1175411415000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__orbitz
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
If a connection is lost while performing an action the current commands errback
is not fired.  For example, downloading a file and the FTP server restarts.  I
do not have a testcase right now (I figure dit out while hacking around a large
project I have) but if one is needed I will try to provide one.  Here is the fix
I came up with, I think it should work as intended without any negative
consequences.

Index: /usr/local/twisted/Twisted/twisted/protocols/ftp.py
===================================================================
--- /usr/local/twisted/Twisted/twisted/protocols/ftp.py (revision 14116)
+++ /usr/local/twisted/Twisted/twisted/protocols/ftp.py (working copy)
@@ -1535,6 +1535,9 @@
             # We're recursing; bail out here for simplicity
             return error
         self._failed = 1
+        if self.nextDeferred:
+            self.nextDeferred.errback(failure.Failure(ConnectionLost('FTP
connection lost', error)))
+            self.nextDeferred = None
         for ftpCommand in self.actionQueue:
             ftpCommand.fail(failure.Failure(ConnectionLost('FTP connection
lost', error)))
         return error
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Test cases are always better :)  The fix close to correct (I am not sure about
the arguments to .errback()), but a test would make it easier to be sure.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
How does one test against a flaky ftpserver in a unit test?

On Jul 14, 2005, at 2:13 PM, Jp Calderone [[Twisted issue tracker]] 
wrote:

>
> Jp Calderone &lt;exarkun@...> added the comment:
>
> Test cases are always better :)  The fix close to correct (I am not 
> sure about
> the arguments to .errback()), but a test would make it easier to be 
> sure.
>
> ----------
> nosy: +exarkun
> status: new -> chatting
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1107>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Don't use the FTP server in the unit test.  You're not trying to test the FTP
server, you want to test the FTP client.

The obvious approach to testing this would be connecting to a dummy server that
has hard-coded responses and then drops its connection... this is similar to how
the testFailedRETR test works.  

More extreme would be to not connect at all, and simply create an FTPClient,
queue some commands, and then call its connectionLost.  In fact, there's already
a "testErrbacksUponDisconnect" in test_ftp that does this!

I wonder why that test doesn't catch your problem...
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
testErrbacksUponDisconnect, I'd imagine the problem is:
         from twisted.internet.main import CONNECTION_LOST
         ftpClient.connectionLost(failure.Failure(CONNECTION_LOST))
         self.failUnless(1, len(m))

Shouldn't it be more like: self.failUnless(len(m)).

I cannot get that test to pass as-is anyways:

[ERROR]: twisted.test.test_ftp.FTPClientTests.testErrbacksUponDisconnect

   File "/usr/local/twisted/imports/twisted/internet/defer.py", line 
360, in _runCallbacks
     self.result = callback(self.result, *args, **kw)
   File "/usr/local/twisted/imports/twisted/protocols/ftp.py", line 
1715, in fail
     self.transport.loseConnection()
exceptions.AttributeError: 'NoneType' object has no attribute 
'loseConnection'

Perhaps I am not running it properly though:
trial twisted.test.test_ftp

I'm not familiar with trial but i'm under the impression that the 
failUnless will always succeed.

On Jul 14, 2005, at 8:19 PM, Andrew Bennetts [[Twisted issue tracker]] 
wrote:

>
> Andrew Bennetts &lt;andrew-twistedbugs@...> added the comment:
>
> Don't use the FTP server in the unit test.  You're not trying to test 
> the FTP
> server, you want to test the FTP client.
>
> The obvious approach to testing this would be connecting to a dummy 
> server that
> has hard-coded responses and then drops its connection... this is 
> similar to how
> the testFailedRETR test works.
>
> More extreme would be to not connect at all, and simply create an 
> FTPClient,
> queue some commands, and then call its connectionLost.  In fact, 
> there's already
> a "testErrbacksUponDisconnect" in test_ftp that does this!
>
> I wonder why that test doesn't catch your problem...
>
> ----------
> topic: +ftp
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1107>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I changed the failUnless line, but the test still passes for me :/
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
How did you run it? I just did trial twisted.test.test_ftp  Is this 
incorrect?  Do i have to specify a special reactor or something that 
set's up a fake transport?

On Jul 16, 2005, at 5:44 PM, Jp Calderone [[Twisted issue tracker]] 
wrote:

>
> Jp Calderone &lt;exarkun@...> added the comment:
>
> I changed the failUnless line, but the test still passes for me :/
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1107>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
Hrm, if I understand FTP correctly, you have a connection that the 
commands go over and then the data connection.  Does this test check to 
see what happens when stuff is going over the data connection and the 
main connection files? Or vice versa?  Does it matter which it tests?

How are you running the test? I still get errors about transport being 
NoneType, any ideas?  If I can get the test running I'll try to make it 
test what happens to me.

On Jul 16, 2005, at 5:44 PM, Jp Calderone [[Twisted issue tracker]] 
wrote:

>
> Jp Calderone &lt;exarkun@...> added the comment:
>
> I changed the failUnless line, but the test still passes for me :/
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1107>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I run the tests in the usual way:

trial twisted.test.test_ftp

And they all pass.  I am running with Python 2.2, 2.3, and 2.4 in Ubuntu.  The
test also passes on all the buildslaves (I committed the failUnless fix).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
I am not much closer to figuring out a fix, however if I take the 
ftpclient.py example and take all of the body of the connectionmade 
function and replace it with a retrieveFile which has a callback and 
errback, then restart the ftpserver in the middle of downloading 
neither the callback or errback are fired.  I do not know why.  The 
patch i submitted earlier makes it so under that condition the errback 
is called, however in the case of a bad login, for instance, the 
errback is not called.

Still trying to figure it out.

On Jul 18, 2005, at 8:30 PM, Jp Calderone [[Twisted issue tracker]] 
wrote:

>
> Jp Calderone &lt;exarkun@...> added the comment:
>
> I run the tests in the usual way:
>
> trial twisted.test.test_ftp
>
> And they all pass.  I am running with Python 2.2, 2.3, and 2.4 in 
> Ubuntu.  The
> test also passes on all the buildslaves (I committed the failUnless 
> fix).
>
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#1107>
> _________________________________________________________
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="orbitz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|orbitz commented|
|-|-|

```
#!html
<pre>
When a connection dies during something like a RETR a DeferredList is created
with several deferreds.  One of them is waiting for some sort of code back from
the FTP server either saying success or failure.  However since the connection
is dead, this code never comes in, but rather connectionLost is called on the
protocol.  Because the waiting deferred is no in self.nextDeferred, _fail only
calls errbacks on the queued commands the DeferredList returned from
retreiveFile never fires because the deferred waiting on the command response is
never fired.  This simply fires it, but because at this point nextDeferred might
already be fired then it needs to be wrapped in a try/except to catch it.  This
seems to work for me.  I can't get trial to run though so I cannot test it fully.



@@ -1731,6 +1732,12 @@
             # We're recursing; bail out here for simplicity
             return error
         self._failed = 1
+       try:
+           if self.nextDeferred:
+               self.nextDeferred.errback(failure.Failure(ConnectionLost('FTP
connection lost', error)))
+               self.nextDeferred = None
+       except defer.AlreadyCalledError:
+           pass
         for ftpCommand in self.actionQueue:
             ftpCommand.fail(failure.Failure(ConnectionLost('FTP connection
lost', error)))
         return error
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
The explanation and patch both sound reasonable.

I'll try to write a test case that triggers the problem on my system.  Thanks!
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [19930]) Merge ftp-client-error-1160+1107

Author: therve
Reviewer: Jerub
Fixes [#7271](https://github.com/twisted/twisted/issues/7271)
Fixes [#11434](https://github.com/twisted/twisted/issues/11434)

Unwrap the FirtError exception from FTPClient failed commands in non-passive
mode, and fire errback of current command (usually happen when losing a
connection during a transfer). Add tests for this situation.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [19931]) Revert r19930: failure in web2 tests.

Refs [#7271](https://github.com/twisted/twisted/issues/7271)
Refs [#11434](https://github.com/twisted/twisted/issues/11434)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set status to `closed`|
|-|-|

(In [19936]) Merge ftp-client-error-1160+1107

Author: therve
Reviewer: jerub
Fixes [#7271](https://github.com/twisted/twisted/issues/7271)
Fixes [#11434](https://github.com/twisted/twisted/issues/11434)

Unwrap the FirtError exception from FTPClient failed commands in non-passive
mode, and fire errback of current command (usually happen when losing a
connection during a transfer). Add tests for this situation, and fix a test.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
